tools/clang/lib/SPIRV/SpirvEmitter.cpp

```diff
@@ -1083,16 +1095,29 @@ void SpirvEmitter::doFunctionDecl(const FunctionDecl *decl) {
                                decl->hasAttr<HLSLPreciseAttr>(), func);
 
   if (spirvOptions.debugInfoRich && decl->hasBody()) {
-    auto line = astContext.getSourceManager().getPresumedLineNumber(loc);
-    auto column = astContext.getSourceManager().getPresumedColumnNumber(loc);
-    auto *source = getRichDebugInfo().source;
-    auto *parentScope = getRichDebugInfo().scopeStack.back();
+    const auto &sm = astContext.getSourceManager();
+    const uint32_t line = sm.getPresumedLineNumber(loc);
+    const uint32_t column = sm.getPresumedColumnNumber(loc);
+
+    const StringRef file = sm.getPresumedLoc(loc).getFilename();
+    auto it = debugInfo.find(file);
+    RichDebugInfo *info = nullptr;
+    if (it == debugInfo.end()) {
+      auto *dbgSrc = spvBuilder.createDebugSource(file);
+      debugInfo[file] =
+          RichDebugInfo(dbgSrc, spvBuilder.createDebugCompilationUnit(dbgSrc));
+      info = &debugInfo.back().second;
+    } else {
+      info = &it->second;
+    }
```

Comment:
Line 758-772 seem the same as line 1098-1112. Maybe we should make a helper function that does all the work (get line column, and info)

Reply:
This was updated later :)